hazelcast-spring/src/main/java/com/hazelcast/spring/transaction/HazelcastTransactionManager.java

```diff
@@ -159,22 +158,41 @@ protected void doRollback(DefaultTransactionStatus status) throws TransactionExc
         HazelcastTransactionObject txObject = (HazelcastTransactionObject) status.getTransaction();
         if (status.isDebug()) {
             logger.debug("Rolling back Hazelcast transaction on TransactionContext ["
-                    + txObject.getTransactionContextHolder().getContext() + "]");
+                    + txObject.getTransactionContextHolder() + "]");
         }
 
-        TransactionContext tx = txObject.getTransactionContextHolder().getContext();
+        TransactionContext tx = txObject.getTransactionContextHolder();
         tx.rollbackTransaction();
     }
 
     @Override
     protected void doCleanupAfterCompletion(Object transaction) {
         HazelcastTransactionObject txObject = (HazelcastTransactionObject) transaction;
-
         if (txObject.isNewTransactionContextHolder()) {
             TransactionSynchronizationManager.unbindResourceIfPossible(hazelcastInstance);
         }
+    }
+
+    @Override
+    protected Object doSuspend(Object transaction) throws TransactionException {
+        HazelcastTransactionObject txObject = (HazelcastTransactionObject) transaction;
+
+        TransactionContext transactionContext = txObject.getTransactionContextHolder();
+        transactionContext.suspendTransaction();
 
-        txObject.getTransactionContextHolder().clear();
```

Comment:
How come we no longer mark the transaction as non-active after it has completed (committed or rolled back)?

Reply:
the `TransactionContextHolder` now implements  `TransactionContext` intarface and `TransactionContextHolder.transactionActive` flag managed within `TransactionContext` methods, so it becomes `false` when transaction commited or rolled back.